### DISCUSSION TITLE
Bulk delete/edit multiple caches

### DISCUSSION THREAD

**Discussion by mohaali482** [arthurfiorette/axios-cache-interceptor#639] [Q&A] [answered]
I was trying to cache different pages of an API, for example, I was trying to cache the API `/users/?limit=10&offset=10` with key `list-users-limit-10-offset-10`. This works fine but the issue came when I was trying to invalidate all the caches that start with the `list-users` key, but I didn't find a way so I was thinking of using a set to store the keys that I was using, and then iterate over the set to remove every cache with the key in that set. But this seems too incorrect to use. Is there a better approach to use? Thank you.

**Comment by arthurfiorette** [comment, marked as answer]
Hey @mohaali482, internally, all request URLs are hashed into a unique ID, making it impossible to do such thing like a `startsWith` comparison, as both URLs will be transformed to completely different numbers.

https://www.npmjs.com/package/object-code

However, you its possible to override your own key generator function to include the original id (only to keep uniqueness) + path (or something that you would use to group requests).

https://axios-cache-interceptor.js.org/guide/request-id#custom-generator

Something like this:

```ts
const axios = setupCache(Axios, {
  keyGenerator: buildKeyGenerator((request) => {
     // Would return something like -2389472895my-group-identifier
     return defaultKeyGenerator(request) + 'my-group-identifier';
  }
});
```

And then, implement your own storage that deletes all keys ending with `'my-group-identifier'`. I'd only recommend you to keep the group identifier last to get faster comparisons with complete unique IDs.

**Reply by mohaali482** [reply]
Thank you for the fast reply.

I was wondering, is there a way to delete the keys without implementing my own storage, just by using the `cache.update` like on the 

https://axios-cache-interceptor.js.org/guide/invalidating-cache#programmatically

I really liked this approach.

**Reply by arthurfiorette** [reply]
Nope, cache update was designed to update keys you already know its existence. The only approach I can think of, for now, would be to override the key generator and add another method on your storage.

**Reply by mohaali482** [reply]
I am really confused on how to add a method on the storage and calling it on mutation, can you reference me to the API, Thank you.

**Reply by arthurfiorette** [reply]
[How is memory storage doing it](https://github.com/arthurfiorette/axios-cache-interceptor/blob/efa7302b2d7687e4907738b05e52d96d25da07ba/src/storage/memory.ts#L103)

**Reply by mohaali482** [reply]
Ok, Thank you

**Reply by arthurfiorette** [reply]
can you mark it as answered?

**Reply by mohaali482** [reply]
I made the following, can you check it out?


```
import { MemoryStorage, buildMemoryStorage } from "axios-cache-interceptor";

export function buildCustomMemoryStorage() {
  let storage: CustomMemoryStorage;

  storage = {
    ...buildMemoryStorage(),

    removeByGroup: (group: string) => {
      const keys = storage.getKeysByGroup(group);
      keys.forEach((key) => storage.remove(key));
    },

    getKeysByGroup: (group: string) => {
      const keys = Object.keys(storage.data);
      return keys.filter((key) => key.endsWith(group));
    },
  };

  return storage;
}

export interface CustomMemoryStorage extends MemoryStorage {
  removeByGroup: (group: string) => void;
  getKeysByGroup: (group: string) => string[];
}

```
